handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -239,34 +239,36 @@ protected void deallocate() {
 
         Lock readerLock = context.ctxLock.readLock();
         readerLock.lock();
-        try {
-            ssl = SSL.newSSL(context.ctx, !context.isClient());
-        } finally {
-            readerLock.unlock();
-        }
-        try {
-            networkBIO = SSL.bioNewByteBuffer(ssl, context.getBioNonApplicationBufferSize());
+        synchronized (this) {
+            try {
+                ssl = SSL.newSSL(context.ctx, !context.isClient());
+            } finally {
+                readerLock.unlock();
```

Comment:
I would prefer to not wrap the readerLock inside the scope of this lock to avoid any potential dependency and deadlock between the two.